src/charm.py

```
@@ -3,113 +3,117 @@
-    def _pebble_layer(self):
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
-            "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {
-                        "GUNICORN_CMD_ARGS": f"--log-level {self.model.config['log-level']}"
-                    },
-                }
-            },
-        }
-
-
-if __name__ == "__main__":  # pragma: nocover
-    ops.main.main(IsCharmsTemplateCharm)
+        self._restart(event)
+
+    def _restart(self, _: ops.EventBase) -> None:
+        """Reconciliation for the jenkins agent charm."""
+        if not self.model.get_relation(AGENT_RELATION):
+            self.model.unit.status = ops.BlockedStatus("Waiting for relation.")
+            return
+
+        if not self.state.agent_relation_credentials:
+            self.model.unit.status = ops.WaitingStatus("Waiting for complete relation data.")
+            logger.info("Waiting for complete relation data.")
+            return
+
+        self.model.unit.status = ops.MaintenanceStatus("Starting agent service.")
+        try:
+            self.jenkins_agent_service.restart()
+        except service.ServiceRestartError as e:
+            logger.debug("Error restarting the agent service %s", e)
+            self.model.unit.status = ops.BlockedStatus("Readiness check failed")
+            return
+
+        # TODO: handle cases where downloading the binary takes a long time
+        self.model.unit.status = ops.WaitingStatus("Waiting for agent service to be up.")
+        if not self.jenkins_agent_service.readiness_check():
+            self.model.unit.status = ops.BlockedStatus("Readiness check failed.")
```

Same here :D

---

I'll just let it error out for now, the hook can be reran to self-correct any errors. Timeout for downloading the binary is currently set at 60s: https://github.com/Thanhphan1147/jenkins-agent-deb/blob/373b0aac22be8f150aa9885ca30be7cbb1b75147/jenkins-agent#L42